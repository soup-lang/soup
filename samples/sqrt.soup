func abs(x)
    if x < 0
        return -x
    return x

func square(x)
    return x * x

func average(x, y)
    return (x + y) / 2

func sqrt(x)
    PRECISION: 0.001
    func is_good_enough(guess)
        return abs(square(guess) - x) < PRECISION
    func improve_guess(guess)
        return average(guess, x / guess)
    func sqrt_iter(guess)
        if is_good_enough(guess)
            return guess
        return sqrt_iter(improve_guess(guess))
    return sqrt_iter(1.0)

console.log(sqrt(10.24))
console.log(sqrt(10.2))
console.log(sqrt(10.0))
