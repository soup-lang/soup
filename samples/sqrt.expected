3.2000000066028047
3.1937438908441074
3.162277665175675
